staker/bold/bold_staker.go

```diff
@@ -0,0 +1,503 @@
+		return nil, fmt.Errorf("could not create state manager: %w", err)
+	}
+	providerHeights := []l2stateprovider.Height{blockChallengeLeafHeight}
+	for i := uint64(0); i < config.NumBigSteps; i++ {
+		providerHeights = append(providerHeights, bigStepHeight)
+	}
+	providerHeights = append(providerHeights, smallStepHeight)
+	provider := l2stateprovider.NewHistoryCommitmentProvider(
+		stateProvider,
+		stateProvider,
+		stateProvider,
+		providerHeights,
+		stateProvider,
+		nil, // Nil API database for the history commitment provider, as it will be provided later. TODO: Improve this dependency injection.
+	)
+	// The interval at which the challenge manager will attempt to post assertions.
+	postingInterval := config.AssertionPostingInterval
+	// The interval at which the manager will scan for newly created assertions onchain.
+	scanningInterval := config.AssertionScanningInterval
+	// The interval at which the manager will attempt to confirm assertions.
+	confirmingInterval := config.AssertionConfirmingInterval
+
+	stackOpts := []challengemanager.StackOpt{
+		challengemanager.StackWithName(config.StateProviderConfig.ValidatorName),
+		challengemanager.StackWithMode(BoldModes[config.Mode]),
+		challengemanager.StackWithPollingInterval(scanningInterval),
+		challengemanager.StackWithPostingInterval(postingInterval),
+		challengemanager.StackWithConfirmationInterval(confirmingInterval),
+		challengemanager.StackWithTrackChallengeParentAssertionHashes(config.TrackChallengeParentAssertionHashes),
+		challengemanager.StackWithHeaderProvider(l1Reader),
+	}
+	if config.API {
+		apiAddr := fmt.Sprintf("%s:%d", config.APIHost, config.APIPort)
+		stackOpts = append(stackOpts, challengemanager.StackWithAPIEnabled(apiAddr, config.APIDBPath))
+	}
+
+	manager, err := challengemanager.NewChallengeStack(
+		assertionChain,
+		provider,
+		stackOpts...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create challenge manager: %w", err)
+	}
+	return manager, nil
+}
+
+// Read the creation info for an assertion by looking up its creation
+// event from the rollup contracts.
+func readBoldAssertionCreationInfo(
+	ctx context.Context,
+	rollup *boldrollup.RollupUserLogic,
+	client bind.ContractFilterer,
+	rollupAddress common.Address,
+	assertionHash common.Hash,
+) (*protocol.AssertionCreatedInfo, error) {
+	var creationBlock uint64
+	var topics [][]common.Hash
+	if assertionHash == (common.Hash{}) {
+		rollupDeploymentBlock, err := rollup.RollupDeploymentBlock(&bind.CallOpts{Context: ctx})
+		if err != nil {
+			return nil, err
+		}
+		if !rollupDeploymentBlock.IsUint64() {
+			return nil, errors.New("rollup deployment block was not a uint64")
+		}
+		creationBlock = rollupDeploymentBlock.Uint64()
+		topics = [][]common.Hash{{assertionCreatedId}}
+	} else {
+		var b [32]byte
+		copy(b[:], assertionHash[:])
+		node, err := rollup.GetAssertion(&bind.CallOpts{Context: ctx}, b)
+		if err != nil {
+			return nil, err
+		}
+		creationBlock = node.CreatedAtBlock
```

Comment:
It is set by the rollup contract as `block.number` when the assertion is created on the parent chain. I presume this will refer to the parent chain block number for L3s

Reply:
No, unfortunately `block.number` on arb1 returns the L1 block number. For the old rollup we had a separate lookup table for this accessible via getNodeCreationBlockForLogLookup. Do we have the same in bold?

Reply:
Looks like we'll have to change a few places inside of the bold/ folder as well to fix this behavior. Added to tasks here https://linear.app/offchain-labs/issue/NIT-2971/bold-l3-support-is-broken-as-assertioncreatedatblock-points-to-the-l1s. It seems L3 support won't work until we fix it

Reply:
Sounds good, since this only blocks L3 support I don't think it needs to block merging